File: .config/ags/utility/weather.ts
import { interval, Variable } from 'astal';
import { ServiceData, ServiceStatus } from './service';
import { location } from './location';
import { fetchJsonAsync } from './network';
import { env } from './environment';

export enum Precipitation {
  None,
  Rain,
  Thunderstorm,
  Snow,
}

export interface Weather {
  temperature: {
    real: number,
    feel: number
  },
  precipitation: Precipitation,
  clouds: number,
  death: boolean
};

export type WeatherData = ServiceData<Weather>;

export const weather = Variable<WeatherData>({ status: ServiceStatus.Unavailable });

const updateWeather = async () => {
  const previousLocation = location.get();

  if (previousLocation.status === ServiceStatus.Unavailable) return;

  let { latitude, longitude } = previousLocation;

  try {
    const data = await fetchJsonAsync(`https://api.openweathermap.org/data/2.5/weather?lat=${latitude}&lon=${longitude}&units=metric&lang=ru&appid=${env.WEATHER_API_KEY}`, 10*1000, 10);

    console.log(data);

    const temperature = {
      real: data.main.temp as number,
      feel: data.main.feels_like as number
    };

    const statuses: string[] = data.weather.map((item: any) => item.id.toFixed());

    let precipitation = Precipitation.None;

    if (statuses.some(status => status.startsWith('2')))
      precipitation = Precipitation.Thunderstorm;
    else if (statuses.some(status => status.startsWith('6')))
      precipitation = Precipitation.Snow;
    else if (statuses.some(status => status.startsWith('3') || status.startsWith('5')))
      precipitation = Precipitation.Rain;

    let death =
      temperature.feel < -25 || temperature.feel > 28 ||
      statuses.some(status => status in ['781', '762', '622', '602', '522', '504', '503', '502', '314', '312', '302', '232', '212', '202'])
    ;

    weather.set({
      status: ServiceStatus.Available,
      temperature,
      precipitation,
      clouds: data.clouds.all,
      death
    });
  } catch {}
};

interval(10*60*1000, updateWeather);
location.subscribe(updateWeather);

File: .config/eww/scripts/weather.py
#!/usr/bin/env python

import requests
import dotenv
import os
import json

try:
  location = requests.get('http://ip-api.com/json/?lang=ru&fields=57552').json()

  dotenv.load_dotenv()
  KEY = os.getenv('WEATHER_API_KEY')

  weather = requests.get(f'https://api.openweathermap.org/data/2.5/weather?lat={location['lat']}&lon={location['lon']}&units=metric&lang=ru&appid={KEY}').json()

  statuses = [str(status['id']) for status in weather['weather']]
  death_statuses = ['781', '762', '622', '602', '522', '504', '503', '502', '314', '312', '302', '232', '212', '202']

  real_temp = weather['main']['temp']
  feel_temp = weather['main']['feels_like']

  result = {
    'available': True,
    'city': location['city'],
    'real_temp': real_temp,
    'feel_temp': feel_temp,
    'death': feel_temp < -25 or feel_temp > 28 or any(status in death_statuses for status in statuses),
    'thunder': any(status.startswith('2') for status in statuses),
    'snow': any(status.startswith('6') for status in statuses),
    'rain': any(status.startswith(('3', '5')) for status in statuses),
    'clouds': weather['clouds']['all'] > 25,
  }

  print(json.dumps(result))

except Exception:
  print(json.dumps({ 'available': False }))
